library/core/src/result.rs

```diff
@@ -230,24 +230,30 @@
 //!
 //! # Representation
 //!
-//! In some cases, [`Result<T, E>`] will gain the same size, alignment, and ABI
-//! guarantees as [`Option<U>`] has. One of either the `T` or `E` type must be a
-//! type that qualifies for the `Option` [representation guarantees][opt-rep],
-//! and the *other* type must meet all of the following conditions:
+//! In some cases, [`Result<T, E>`] comes with size, alignment, and ABI guarantees.
+//! Specifically, one of either the `T` or `E` type must be a type that qualifies for the `Option`
+//! [representation guarantees][opt-rep] (let's call that type `I`), and the *other* type must meet
+//! all of the following conditions:
 //! * Is a zero-sized type with alignment 1 (a "1-ZST").
-//! * Has no fields.
+//! * Is either a struct/tuple with no fields, or an enum with no variants.
```

Comment:
@theemathas argues that this is a breaking change -- arguably, the `!` type, or `[T; 0]`, or enums with a single variant and no fields were allowed by the previous wording.

The new wording matches what actually got implemented in https://github.com/rust-lang/rust/pull/122253. That PR also used the "has no fields" wording in a doc comment, which (as it turns out) doesn't actually match what is implemented. @Lokathor then copied it to the docs in https://github.com/rust-lang/rust/pull/124870, and later I copied that wording into the FCP [here](https://github.com/rust-lang/rust/pull/130628#issuecomment-2365492463).

Reply:
It is also worth mentioning that the [ABI docs](https://doc.rust-lang.org/nightly/std/primitive.fn.html#abi-compatibility) actually document a stronger guarantee: if someone defines their own `enum Result`, they get this guarantee whenever the other type is a 1-ZST, without any further caveats.

This is based on the FCP [here](https://github.com/rust-lang/rust/pull/130628#issuecomment-2402761599). The Result docs never got updated to match that. Should they? And what about the lint? The reason the lint works like this is semver; the ABI docs are deliberately not semver-aware since one may want to use them within one's own crate.

Reply:
If the compiler supports/checks the more permissive form given in the FCP, then I think the docs should match.

Reply:
(and if the compiler doesn't support the FCP's wording then the compiler should be updated, followed by a docs update)

Reply:
> Real Question: is an array a struct/tuple, or is it an enum?

Neither.

@traviscross so the real question here is this one I think:
> The Result docs should either match the FFI-types lint or the ABI guarantees, and currently they do neither, so we definitely want to change something here.

Seems like you're leaning towards having it match the ABI guarantees, so I'll update the PR for that. (I won't be updating the FFI-types lint here, that lint needs a redesign since it's currently not even clear when and how it should take semver into consideration, among other issues.)

Reply:
If an array is neither then are we deliberately excluding 0-sized arrays from these docs? Or should the wording also permit them?

Reply:
The FFI lint considers 0-zised 1-aligned arrays to not qualify for the guarantee. I cannot tell whether that is deliberate or not.

Reply:
Oh, yeah, I was looking in comments for the lint it seems. No problem then, I guess.